### PR TITLE
[SPARK-13704][CORE][YARN] Reduce rack resolution time

### PR DESCRIPTION
**Pull request by squito**
## What changes were proposed in this pull request?

When you submit a stage on a large cluster, rack resolving takes a long time when initializing TaskSetManager because a script is invoked to resolve the rack of each host, one by one.
Based on current implementation, it takes 30~40 seconds to resolve the racks in our 5000 nodes' cluster. After applied the patch, it decreased to less than 15 seconds.

YARN-9332 has added an interface to handle multiple hosts in one invocation to save time. But before upgrading to the newest Hadoop, we could construct the same tool in Spark to resolve this issue.

## How was this patch tested?

UT and manually testing on a 5000 node cluster.

### CODE REVIEW

**Comment by squito** [comment]
This is an update of https://github.com/apache/spark/pull/23951, to address the last couple review items.  All credit still to @LantaoJin.  Note that I also removed the skipRackResolution option entirely, rather than moving it to a new conf, because I think this should improve time enough to make that unnecessary.  But I haven't tested on a 5k node cluster :)

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnRMClient.scala [hunk 1]
```diff
@@ -83,7 +83,7 @@ private[spark] class YarnRMClient extends Logging {
       localResources: Map[String, LocalResource]): YarnAllocator = {
     require(registered, "Must register AM before creating allocator.")
     new YarnAllocator(driverUrl, driverRef, conf, sparkConf, amClient, appAttemptId, securityMgr,
-      localResources, new SparkRackResolver())
+      localResources, new SparkRackResolver(conf))
```
Comment (vanzin) [inline resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnRMClient.scala]: Do you want the shared instance here instead?

File: resource-managers/yarn/src/main/scala/org/apache/spark/scheduler/cluster/YarnScheduler.scala [hunk 2]
```diff
@@ -17,23 +17,29 @@
...
 
 private[spark] class YarnScheduler(sc: SparkContext) extends TaskSchedulerImpl(sc) {
 
-  // RackResolver logs an INFO message whenever it resolves a rack, which is way too often.
-  if (Logger.getLogger(classOf[RackResolver]).getLevel == null) {
-    Logger.getLogger(classOf[RackResolver]).setLevel(Level.WARN)
-  }
+  override val defaultRackValue: Option[String] = Some(NetworkTopology.DEFAULT_RACK)
+
+  private[spark] val resolver = new SparkRackResolver(sc.hadoopConfiguration)
```
Comment (vanzin) [inline resource-managers/yarn/src/main/scala/org/apache/spark/scheduler/cluster/YarnScheduler.scala]: Use the shared instance?

Or, if not using the shared instance, then the `SparkRackResolver` object can go away.

**Review by vanzin — COMMENTED** [review]
Looks ok but there's a conflict...

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnRMClient.scala (hunk 1 above)
Comment (squito) [reply resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnRMClient.scala]: oh good point, i do want the shared instance, to use a shared `CachedDnsToSwitchMapping`.  In cluster mode that will make sure we re-use the cache.

**Comment by heary-cao** [comment]
retest this please

**Review by attilapiros — COMMENTED** [review]
I have found only some really minor things.

File: core/src/test/scala/org/apache/spark/scheduler/TaskSetManagerSuite.scala [hunk 3]
```diff
@@ -1602,4 +1612,40 @@ class TaskSetManagerSuite extends SparkFunSuite with LocalSparkContext with Logg
...
+        // we'll reject them at NODE_LOCAL level,
+        // but accept them at RACK_LOCAL level if they're on OK racks
+        val hostIdx = 100 + rackIdx
+        assert(manager.resourceOffer("exec" + hostIdx, "host" + hostIdx, TaskLocality.NODE_LOCAL)
+          .isEmpty)
+        assert(manager.resourceOffer("exec" + hostIdx, "host" + hostIdx, TaskLocality.RACK_LOCAL)
+          .isDefined)
+      }
+    }
+    assert(FakeRackUtil.numBatchInvocation === 1)
```
Comment (attilapiros) [inline core/src/test/scala/org/apache/spark/scheduler/TaskSetManagerSuite.scala]: As `addPendingTasks()` is called during the construction of `TaskSetManager` instance (and that is the only one place where multiple hosts can be passed to `getRacksForHosts()`) I did not get why `numBatchInvocation === 1` is so important that it is emphasised by this assert. 

I assume we are thinking about any potential future code which would use the `getRacksForHosts()` (meanwhile it is an expensive call), am I right?

File: core/src/test/scala/org/apache/spark/scheduler/TaskSetManagerSuite.scala [hunk 4]
```diff
@@ -1631,7 +1635,14 @@ class TaskSetManagerSuite extends SparkFunSuite with LocalSparkContext with Logg
     }
     val taskSet = FakeTask.createTaskSet(100, locations: _*)
     val clock = new ManualClock
+    // make sure we only do one rack resolution call, for the entire batch of hosts, as this
+    // can be expensive.  the FakeTaskScheduler calls rack resolution more than the real one
```
Comment (attilapiros) [inline core/src/test/scala/org/apache/spark/scheduler/TaskSetManagerSuite.scala]: Nit: "expensive.  the " =>  "expensive. The "

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/SparkRackResolver.scala [hunk 5]
```diff
@@ -17,24 +17,105 @@
...
+ * With that, we could just directly use [[org.apache.hadoop.yarn.util.RackResolver]].
+ * In the meantime, this is a re-implementation for spark's use.
+ */
+object SparkRackResolver extends Logging {
+  @volatile private var instance: SparkRackResolver = _
+
+  /**
+   * It will return the static resolver instance.  If there is already an instance, the passed
+   * conf is entirely ignored.  If there is not a shared instance, it will create one with the
+   * given conf.
```
Comment (LantaoJin) [inline resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/SparkRackResolver.scala]: Should we explain how to instantiate a separate resolver with a separate config here?
> Instantiate a separate resolver with a separate config by `new SparkRackResolver(conf)`

Comment (squito) [reply resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/SparkRackResolver.scala]: its kinda obvious, no?

File: core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala [hunk 6]
```diff
@@ -186,8 +186,23 @@ private[spark] class TaskSetManager(
...
+  addPendingTasks()
+
+  private def addPendingTasks(): Unit = {
+    val (_, duration) = Utils.timeTakenMs {
+      for (i <- (0 until numTasks).reverse) {
+        addPendingTask(i, resolveRacks = false)
+      }
+      // Resolve the rack for each host. This can be slow, so de-dupe the list of hosts,
+      // and assign the rack to all relevant task indices.
+      val racks = sched.getRacksForHosts(pendingTasksForHost.keySet.toSeq)
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala]: There's an implicit assumption here that `map.keySet` and `map.values` iterate in the same order. I'm not sure if that's guaranteed, and at the same time I don't see why that wouldn't be the case, but just wanted to point this out.

File: core/src/test/scala/org/apache/spark/scheduler/TaskSetManagerSuite.scala [hunk 7]
```diff
@@ -1323,7 +1337,7 @@ class TaskSetManagerSuite extends SparkFunSuite with LocalSparkContext with Logg
     val e = new ExceptionFailure("a", "b", Array(), "c", None)
     taskSetManagerSpy.handleFailedTask(taskDesc.get.taskId, TaskState.FAILED, e)
 
-    verify(taskSetManagerSpy, times(1)).addPendingTask(anyInt())
+    verify(taskSetManagerSpy, times(1)).addPendingTask(meq(0), meq(false))
```
Comment (vanzin) [inline core/src/test/scala/org/apache/spark/scheduler/TaskSetManagerSuite.scala]: You can actually omit the `meq` in this case.

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/SparkRackResolver.scala [hunk 8]
```diff
@@ -17,24 +17,105 @@
...
+    nodes.toList
+  }
+}
+
+/**
+ * Utility to resolve the rack for hosts in an efficient manner.
+ * It will cache the rack for individual hosts to avoid
+ * repeatedly performing the same expensive lookup.
+ *
+ * Its logic refers [[org.apache.hadoop.yarn.util.RackResolver]] and enhanced.
```
Comment (vanzin) [inline resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/SparkRackResolver.scala]: This paragraph actually refers to the code in the class, now, not the object anymore.

**Review by vanzin — COMMENTED** [review]
Looks good aside from some minor comments. The comments could use some grammar updates too, but not a big deal since you didn't write them.

**Comment by LantaoJin** [comment]
retest this please

File: core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala (hunk 6 above)
Comment (attilapiros) [reply core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala]: I had the exact same thought when I reached that line. 
Even thought about a possible solutions: 

- creating a new val with the value `racks.entrySet` and generating the keys and values from this entry set (as in entry set the key and value is bound together the ordering will be fixed; even with one iteration the key and the value can be generated). 
- Another possible and more elegant solution is calling `racks.asScala.unzip`. 

Both solutions has some performance cost.

Comment (attilapiros) [reply core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala]: Correction for last sentence: the first solution can be done without performance cost but probably the code will be a bit less elegant.

Comment (squito) [reply core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala]: great point, I have updated this to use `unzip`

**Review by squito — COMMENTED** [review]
Thanks @attilapiros @vanzin .  I also updated went through the comments and updated them a bit (I just deleted a few that I thought weren't really adding anything).

**Comment by attilapiros** [comment]
I have checked the new changes (the last commit) and it looks good to me. 
Even run the new unittest as I was a little worried about the matcher-less mockito verify line but it works just fine.

So pending tests otherwise LGTM.

**Comment by LantaoJin** [comment]
retest this please

**Comment by squito** [comment]
The pyspark failures are very strange, I'm not sure if its an infra issue or something else.  I filed https://issues.apache.org/jira/browse/SPARK-27389

**Comment by vanzin** [comment]
Looks fine. I still think that `getRackForHosts` could just return a `Seq[String]` and avoid using `Option` at all - the code in `SparkRackResolver` doesn't seem to ever return a `null` location, but not worth delaying this patch more.

**Comment by LantaoJin** [comment]
> Looks fine. I still think that `getRackForHosts` could just return a `Seq[String]` and avoid using `Option` at all - the code in `SparkRackResolver` doesn't seem to ever return a `null` location, but not worth delaying this patch more.

That's because if `getRacksForHosts` return Seq[String], `getRackForHost` will return a String. If it is like this, the default value of rack will be difficult to set. `None` is the best one.

**Comment by LantaoJin** [comment]
This update PR, LGTM now

**Comment by vanzin** [comment]
> That's because if getRacksForHosts return Seq[String], getRackForHost will return a String. 

If only you could write code to wrap a String in an Option...

**Review by xuanyuanking — APPROVED** [review]
I want to give a +1 for this, seems a great performance optimization on large clusters.

**Comment by squito** [comment]
merged to master.  Thanks @LantaoJin !

**Pull request closed without merging**